L043 refinement: Add autofix for common use of CASE to fill NULL values.

jpy-git opened the pull request:
<!--Firstly, thanks for adding this feature! Secondly, please check the key steps against the checklist below to make your contribution easy to merge.-->

<!--Please give the Pull Request a meaningful title (including the dialect this PR is for if it is dialect specific), as this will automatically be added to the release notes, and then the Change Log.-->

### Brief summary of the change made
<!--If there is an open issue for this, then please include `fixes #XXXX` or `closes #XXXX` replacing `XXXX` with the issue number and it will automatically close the issue when the pull request is merged. Alternatively if not fully closed you can say `makes progress on #XXXX` to create a link on that issue without closing it.-->
This PR enhances L043 to fix #1797. I have also tidied up the docstring and added the corresponding unit tests.

### Are there any other side effects of this change that we should be aware of?
L043 should now flag and fix more cases of unnecessary use of CASE. 

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

jpy-git commented:
@barrywhart Ready for review 😄

barrywhart commented on src/sqlfluff/rules/L043.py:
```diff
@@ -1,154 +1,261 @@
+                coalesce_arg_1 = condition_expression
+                coalesce_arg_2 = KeywordSegment("false")
+                coalesce_arg_1_idx = condition_expression_idx
+                preceding_not = True if then_expression.raw_upper == "FALSE" else False
```
```suggestion
                preceding_not = (then_expression.raw_upper == "FALSE")
```

jpy-git replied:
@barrywhart 🤦 wow I definitely had a brain lapse there, thanks!

barrywhart reviewed (approved):
Looks good to me.

This rule has always been one of the scarier ones to me, because I worry about the risk of "false positives", like if it triggers for a corner case we didn't consider. For example, I think an early version of this rule assumed (incorrectly) that if a case statement used booleans and an else, that one boolean would be false and the other true. Can you think of any potential "gotchas" with these changes, and perhaps we should add more tests to catch them?

jpy-git commented:
@barrywhart so I added some checks to make sure we only operate on CASE statements with a single WHEN and also added a check that the IS NULL condition is the only condition. Iet me make some extra unit tests for cases we think should pass this lint unaffected for extra peace of mind 👍

jpy-git commented:
@barrywhart ive added additional unit test to validate my checks in the code so this should be good to go 🚀
